README.md
wifi_ddwrt[![Build Status](https://travis-ci.com/ros-drivers/wifi_ddwrt?branch=hydro-devel)](https://travis-ci.com/ros-drivers/wifi_ddwrt)
==========

The main documentation for the wifi_ddwrt package is the [wifi_ddwrt](http://wiki.ros.org/wifi_ddwrt) page on the ROS wiki.

For Hydro and newer versions of ROS, use the hydro-devel branch:

    git clone -b hydro-devel https://github.com/ros-drivers/wifi_ddwrt.git

For Groovy and older version of ROS, use the master branch:

    git clone -b master https://github.com/ros-drivers/wifi_ddwrt.git


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package wifi_ddwrt
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.2.1 (2021-05-15)
------------------
* Merge pull request `#15 <https://github.com/ros-drivers/wifi_ddwrt/issues/15>`_ from k-okada/add_travis
  add travsi CI
* update CHANGELOG.rst
* Merge pull request `#14 <https://github.com/ros-drivers/wifi_ddwrt/issues/14>`_ from PR2-prime/noetic-devel
* add travsi CI
* updated for focal/noetic/python3 compatibility
* Merge pull request `#7 <https://github.com/ros-drivers/wifi_ddwrt/issues/7>`_ from furushchev/fix-typo
  [wifi_ddwrt] fix typo
* [wifi_ddwrt] fix typo
* Update README.md
* Create README. Fixes `#1 <https://github.com/ros-drivers/wifi_ddwrt/issues/1>`_
* Contributors: Austin, Dave Feil-Seifer, Kei Okada, Yuki Furuta

* updated for focal/noetic/python3 compatibility (`#14 <https://github.com/ros-drivers/wifi_ddwrt/issues/14>`_)
* [wifi_ddwrt] fix typo (`#7 <https://github.com/ros-drivers/wifi_ddwrt/issues/7>`_)
* Create README. Fixes `#1 <https://github.com/ros-drivers/wifi_ddwrt/issues/1>`_
* Contributors: Austin Hendrix, Dave Feil-Seifer, Yuki Furuta

0.2.2 (2021-05-15)
------------------
* 0.2.1
* update changelogs
* Merge pull request `#15 <https://github.com/ros-drivers/wifi_ddwrt/issues/15>`_ from k-okada/add_travis
  add travsi CI
* update CHANGELOG.rst
* Merge pull request `#14 <https://github.com/ros-drivers/wifi_ddwrt/issues/14>`_ from PR2-prime/noetic-devel
* add travsi CI
* updated for focal/noetic/python3 compatibility
* Merge pull request `#7 <https://github.com/ros-drivers/wifi_ddwrt/issues/7>`_ from furushchev/fix-typo
  [wifi_ddwrt] fix typo
* [wifi_ddwrt] fix typo
* Update README.md
* Create README. Fixes `#1 <https://github.com/ros-drivers/wifi_ddwrt/issues/1>`_
* Contributors: Austin, Dave Feil-Seifer, Kei Okada, Yuki Furuta

0.2.0 (2013-06-03)
------------------
* Un-stack.
* Catkinized!
* Remove references to roslib.
* Remove unused options.
* Remove old rosbuild files.
* removing rosdep.yaml
* bumped version for 0.1.5 release
* fix typo, ticket `#5078 <https://github.com/ros-drivers/wifi_ddwrt/issues/5078>`_
* update stack dependencies
* getting ready for 0.1.4 release
* remove deprecated imports
* adding ros_comm depend
* Only print warning when state changes, not every time
* Added Ubuntu platform tags to manifest
* updating to release 0.1.3
* Catch all exception types and make sure we guard fetchSiteSurvey too
* Catch exceptions in fetchCurrentAP, log them, and continue running
* Updating wifi_drivers to 0.1.2
* Took out the gc.collect from each iteration, as it does not seem to be necessary.
* Fixed ticket texas`#135 <https://github.com/ros-drivers/wifi_ddwrt/issues/135>`_ by not reusing the mechanize.Browser, and by manually clearing uncollectable dictionnaries that get left on the heap.
* updating to 0.1.1
* don't try to create an access point object unless we have active data, fixing `#60 <https://github.com/ros-drivers/wifi_ddwrt/issues/60>`_
* wifi_drivers: stack manifest update
* wifi_drivers: Makefile
* wifi_drivers requires python_mechanize rosdep
* preparing wifi_drivers 0.1.0 into tick-tock
* staging wifi_drivers into tick-tock
* wifi_drivers: updating to newer rosbuild commands
* updating to release 0.1.0
* Add wifi_wrt
* Contributors: Austin Hendrix, blaise, gerkey, jamesb, jtyler, kwc, leibs, wheeler, wim


scripts/ddwrt_config.py

import os, sys
import urllib2, urllib
import base64
import StringIO
import string
import csv

class DDWRT:
    def __init__(self, hostname, username, password):
        self.hostname = hostname
        self.username = username
        self.password = password

    def fetch(self, url, args=None):
        req = urllib2.Request(url)
        base64string = base64.encodestring("%s:%s" % (self.username, self.password))[:-1]

        req.add_header("Authorization", "Basic %s" % base64string)
        if args:
            req.data = urllib.urlencode(args)

        fp = urllib2.urlopen(req)
        body = fp.read()
        fp.close()

        return body

    def connect(self, essid):
        url = "http://%s/apply.cgi" % self.hostname

        args = {}
        args["submit_button"]="Wireless_Basic"
        args["action"]="ApplyTake"
        args["change_action"]="gozila_cgi"
        args["submit_type"]="save"
        args["wl0_nctrlsb"]=""
        args["wl1_nctrlsb"]=""
        args["iface"]=""
        args["wl0_mode"]="sta"
        args["wl0_net_mode"]="n-only"
        args["wl0_ssid"]=essid
        args["wl0_distance"]="500"
        args["eth1_bridged"]="1"
        args["eth1_multicast"]="0"
        args["eth1_nat"]="1"
        args["eth1_ipaddr"]="4"
        args["eth1_ipaddr_0"]="0"
        args["eth1_ipaddr_1"]="0"
        args["eth1_ipaddr_2"]="0"
        args["eth1_ipaddr_3"]="0"
        args["eth1_netmask"]="4"
        args["eth1_netmask_0"]="0"
        args["eth1_netmask_1"]="0"
        args["eth1_netmask_2"]="0"
        args["eth1_netmask_3"]="0"
        args["wl1_mode"]="ap"
        args["wl1_net_mode"]="disabled"
        args["wl1_ssid"]="dd-wrt"
        args["wl1_closed"]="0"
        args["wl1_distance"]="2000"
        args["eth2_bridged"]="1"
        args["eth2_multicast"]="0"
        args["eth2_nat"]="1"
        args["eth2_ipaddr"]="4"
        args["eth2_ipaddr_0"]="0"
        args["eth2_ipaddr_1"]="0"
        args["eth2_ipaddr_2"]="0"
        args["eth2_ipaddr_3"]="0"
        args["eth2_netmask"]="4"
        args["eth2_netmask_0"]="0"
        args["eth2_netmask_1"]="0"
        args["eth2_netmask_2"]="0"
        args["eth2_netmask_3"]="0"

        body = self.fetch(url, args)

    def current_ap(self):
        url = "http://%s/Status_Wireless.live.asp" % self.hostname
        body = self.fetch(url)
        
        line = None
        lines = body.split("\n")

        d = {}
        for line in lines:
          line = line[1:-1]
          line = line.replace("&nbsp;", "")
          parts = line.split("::", 1)
          if len(parts) == 2:
            d[parts[0]] = parts[1]

        essid = d.get('wl_ssid', '')
        wl_channel = d.get('wl_channel', '').split()[0]
        channel = int(wl_channel)
        rate = d.get('wl_rate', '')

        signal = None
        noise = None
        snr = None
        quality = None

        tx_power = d.get('wl_xmit', '')

        active_wireless = d.get('active_wireless', None)
        if active_wireless:
          active_wireless = active_wireless.replace("'", "")
          parts = active_wireless.split(",")
          macaddr = parts[0]
          interface = parts[1]
          if len(parts) == 7:
            signal = int(parts[4])
            noise = int(parts[5])
            snr = int(parts[6])
            quality = signal * 1.24 + 116
          else:
            signal = int(parts[5])
            noise = int(parts[6])
            snr = int(parts[7])
            quality = int(parts[8])/10

        print essid

    def site_survey(self):
        url = "http://%s/Site_Survey.asp" % self.hostname

        body = self.fetch(url)

        lines = body.split("\n")
        for i in range(len(lines)):
          if lines[i].startswith("var table = "):
            break

        aplines = []
        for j in range(i+1, len(lines)):
          if lines[j].startswith(");"): break
          line = lines[j].strip()
          if not line: continue
          if line[0] == ",": line = line[1:]

          aplines.append(line)

        aps = {}
        
        fp = StringIO.StringIO(string.join(aplines, '\n'))
        reader = csv.reader(fp)
        for row in reader:
          essid = row[0]
          macattr = row[2]
          channel = int(row[3])
          rssi = int(row[4])
          noise = int(row[5])
          beacon = int(row[6])

          aps[essid] = row

        for ap, row in aps.items():
            print ap, row[8][:20]

import getopt
def main(argv, stdout, environ):
    progname = sys.argv[0]
    optlist, args = getopt.getopt(argv[1:], "", ["help"])
    
    dd = DDWRT("192.168.1.1", "root", "admin")

    if len(args) == 0:
        dd.current_ap()
        return
        
    cmd = args[0]
    if cmd == "connect":
        dd.connect(args[1])
    elif cmd == "survey":
        dd.site_survey()

if __name__ == "__main__":
    main(sys.argv, sys.stdout, os.environ)



wifi_ddwrt/scripts/ddwrt_sim.py
#!/usr/bin/env python
# Software License Agreement (BSD License)
#
# Copyright (c) 2008, Willow Garage, Inc.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
#  * Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above
#    copyright notice, this list of conditions and the following
#    disclaimer in the documentation and/or other materials provided
#    with the distribution.
#  * Neither the name of the Willow Garage nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.
#
# Revision $Id: gossipbot.py 1013 2008-05-21 01:08:56Z sfkwc $

import os, sys, string, time, getopt, re
import StringIO
import random

import rospy
from wifi_ddwrt.msg import *
from pr2_msgs.msg import AccessPoint

from std_msgs.msg import Header

wifi_data = {"header": {"seq": 24, "stamp": 1257901832.6, "frame_id": "0"}, "networks": [{"macaddr": "00:18:F8:F9:6C:41", "essid": "willow", "channel": 11, "rssi": -35, "noise": -88, "beacon": 100}, {"macaddr": "00:18:F8:F9:6B:BD", "essid": "willow", "channel": 11, "rssi": -71, "noise": -88, "beacon": 100}, {"macaddr": "00:18:F8:F9:6C:1D", "essid": "willow", "channel": 1, "rssi": -70, "noise": -92, "beacon": 100}, {"macaddr": "00:30:44:03:1F:F9", "essid": "PRLAN", "channel": 1, "rssi": -73, "noise": -92, "beacon": 100}, {"macaddr": "00:30:44:03:1F:F4", "essid": "PRGLAN", "channel": 2, "rssi": -80, "noise": -89, "beacon": 100}, {"macaddr": "00:18:F8:F9:6C:4D", "essid": "willow", "channel": 6, "rssi": -81, "noise": -86, "beacon": 100}, {"macaddr": "00:18:F8:F9:6C:44", "essid": "willow", "channel": 6, "rssi": -79, "noise": -85, "beacon": 100}]}

def minmax(v, lower, upper):
    if v < lower: v = lower
    if v > upper: v = upper
    return v

class WifiAP:
  def __init__(self):
      self.ap = AccessPoint()
      self.t = time.time()
      self._pick_ap()
      
  def _pick_ap(self):
      ap = random.choice(wifi_data['networks'])
      self.ap.macaddr = ap['macaddr']
      self.ap.channel = ap['channel']
      self.ap.essid = ap['essid']
      self.ap.rate = '54 Mbps'
      self.ap.tx_power = '71 mW'
      self.ap.signal = random.randint(-100, -48)
      self.ap.snr = 44
      self.ap.noise = -92
      self.ap.quality = 56

  def _gen_snr(self):
      d = 2
      if time.time() - self.t > 20:
          d = 10
      self.ap.signal = minmax(self.ap.signal + random.randint(-d, d), -70, -20)

      self.ap.snr = minmax(self.ap.snr + random.randint(-d, d), 30, 90)
      self.ap.noise = minmax(self.ap.noise + random.randint(-d, d), -95, -30)
      self.ap.quality = int(self.ap.signal * 1.24 + 116)
  
  def fetchSiteSurvey(self):
    header = Header()
    header.stamp = rospy.Time.now()
    networks = []
    survey = SiteSurvey(header, networks)
    
    for network in wifi_data["networks"]:
      network = Network(network['macaddr'], network['essid'], network['channel'], network['rssi'], network['noise'], network['beacon'])
      survey.networks.append(network)
    return survey

  def fetchCurrentAP(self):
    if time.time() - self.t > 60:
        self._pick_ap()
    self._gen_snr()
    
    #make sure that we put a stamp on things
    self.ap.header = Header()
    self.ap.header.stamp = rospy.Time.now()

    return self.ap

def loop():
  rospy.init_node("wifi_ddwrt")

  ap = WifiAP()

  pub1 = rospy.Publisher("ddwrt/sitesurvey", SiteSurvey)
  pub2 = rospy.Publisher("ddwrt/accesspoint", AccessPoint)

  while not rospy.is_shutdown():
    survey = ap.fetchSiteSurvey()
    pub1.publish(survey)
    node = ap.fetchCurrentAP()
    pub2.publish(node)
    time.sleep(5)
        
def usage(progname):
  print __doc__ % vars()

def main(argv, stdout, environ):
  progname = argv[0]
  optlist, args = getopt.getopt(argv[1:], "", ["help",])

  for (field, val) in optlist:
    if field == "--help":
      usage(progname)
      return

  loop()

if __name__ == "__main__":
  main(sys.argv, sys.stdout, os.environ)
        



nodes/ddwrt.py
#!/usr/bin/env python
# Software License Agreement (BSD License)
#
# Copyright (c) 2008, Willow Garage, Inc.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
#  * Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above
#    copyright notice, this list of conditions and the following
#    disclaimer in the documentation and/or other materials provided
#    with the distribution.
#  * Neither the name of the Willow Garage nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.
#
# Revision $Id: gossipbot.py 1013 2008-05-21 01:08:56Z sfkwc $

import os, sys, string, time, getopt, re
import StringIO

import rospy
from wifi_ddwrt.msg import *
from pr2_msgs.msg import AccessPoint

from mechanize import Browser
from std_msgs.msg import Header
import csv

import gc

def breakUpTrash():
    for item in gc.garbage:
        if type(item) == dict:
            for s in item.keys_iter():
                del item[s]
        del gc.garbage[:]

class WifiAP:
  def __init__(self, hostname, username, password):
    self.hostname = hostname
    self.username = username
    self.password = password

  def newBrowser(self):
    # Create new browsers all the time because its data structures grow
    # unboundedly (texas#135)
    br = Browser()
    br.add_password(self.hostname, self.username, self.password)
    br.set_handle_robots(None)
    return br

  def fetchSiteSurvey(self):
    url = "http://%s/Site_Survey.asp" % self.hostname

    response = self.newBrowser().open(url)

    body = response.read()
    
    #make sure that we put a stamp on things
    header = Header()
    header.stamp = rospy.Time.now()
    networks = []
    survey = SiteSurvey(header, networks)

    lines = body.split("\n")
    for i in range(len(lines)):
      if lines[i].startswith("var table = "):
        break

    aplines = []
    for j in range(i+1, len(lines)):
      if lines[j].startswith(");"): break
      line = lines[j].strip()
      if not line: continue
      if line[0] == ",": line = line[1:]

      aplines.append(line)

    fp = StringIO.StringIO(string.join(aplines, '\n'))
    reader = csv.reader(fp)
    for row in reader:
      essid = row[0]
      macattr = row[2]
      channel = int(row[3].split(' ')[0])
      rssi = int(row[4])
      noise = int(row[5])
      beacon = int(row[6])

      network = Network(macattr, essid, channel, rssi, noise, beacon)
      survey.networks.append(network)
    return survey

  def fetchBandwidthStats(self, interface):
    url = "http://%s/fetchif.cgi?%s" % (self.hostname, interface)
    response = self.newBrowser().open(url)
    body = response.read()

    lines = body.split("\n")

    if len(lines) > 1:
      line = lines[1].strip()
      iparts = line.split(":", 1)
      parts = iparts[1].split()
      print (interface, parts)
      

  def fetchCurrentAP(self):
    url = "http://%s/Status_Wireless.live.asp" % self.hostname
    response = self.newBrowser().open(url)
    body = response.read()

    line = None
    lines = body.split("\n")

    d = {}
    for line in lines:
      line = line[1:-1]
      line = line.replace("&nbsp;", "")
      parts = line.split("::", 1)
      if len(parts) == 2:
        d[parts[0]] = parts[1]
      
    essid = d.get('wl_ssid', '')
    wl_channel = d.get('wl_channel', '').split()[0]
    channel = int(wl_channel)
    rate = d.get('wl_rate', '')
    
    signal = None
    noise = None
    snr = None
    quality = None

    tx_power = d.get('wl_xmit', '')

    active_wireless = d.get('active_wireless', None)
    ap = None
    if active_wireless:
      active_wireless = active_wireless.replace("'", "")
      parts = active_wireless.split(",")
      macaddr = parts[0]
      interface = parts[1]
      if len(parts) == 7:
        signal = int(parts[4])
        noise = int(parts[5])
        snr = int(parts[6])
        quality = signal * 1.24 + 116
      else:
        signal = int(parts[5])
        noise = int(parts[6])
        snr = int(parts[7])
        quality = int(parts[8])/10
      
      #self.fetchBandwidthStats(interface)

      #make sure that we put a stamp on things
      header = Header()
      header.stamp = rospy.Time.now()

      ap = AccessPoint(header=header,
                       essid=essid,
                       macaddr=macaddr,
                       signal=signal,
                       noise=noise,
                       snr=snr,
                       channel=channel,
                       rate=rate,
                       quality=quality,
                       tx_power=tx_power)

    return ap


def loop():
  rospy.init_node("wifi_ddwrt")

  router_ip = rospy.get_param('~router', 'wifi-router')
  username = rospy.get_param('~username', 'root')
  password = rospy.get_param('~password', '')

  ap = WifiAP(router_ip, username, password)

  pub1 = rospy.Publisher("ddwrt/sitesurvey", SiteSurvey)
  pub2 = rospy.Publisher("ddwrt/accesspoint", AccessPoint)

  r = rospy.Rate(.5)
  lastTime = 0
  last_ex = ''
  while not rospy.is_shutdown():
    breakUpTrash() # Needed because mechanize leaves data structures that the GC sees as uncollectable (texas#135)
    try:
      if time.time() - lastTime > 60:
        survey = ap.fetchSiteSurvey()
        pub1.publish(survey)
        lastTime = time.time()
      node = ap.fetchCurrentAP()
      if node: pub2.publish(node)
      last_ex = ''
    except Exception as e:
      if e != last_ex:
          rospy.logwarn("Caught exception %s" % e)
      last_ex = e
    r.sleep()
        
def test():
  router_ip = rospy.get_param('~router_ip', 'wifi-router')
  username = rospy.get_param('~username', 'root')
  password = rospy.get_param('~password', '')

  ap = WifiAP(router_ip, username, password)
  while 1:
    if 0:
      survey = ap.fetchSiteSurvey()
      print (survey)
    if 1:
      node = ap.fetchCurrentAP()
      print (node)

def usage(progname):
  print (__doc__ % vars())

def main(argv, stdout, environ):
  progname = argv[0]
  optlist, args = getopt.getopt(argv[1:], "", ["help", "test",])

  testflag = 0

  for (field, val) in optlist:
    if field == "--help":
      usage(progname)
      return
    elif field == "--test":
      testflag = 1

  if testflag:
    test()
    return

  loop()

if __name__ == "__main__":
  main(sys.argv, sys.stdout, os.environ)
        

